### ISSUE TITLE
Assign codes to errors

### ISSUE THREAD

**Issue by evrifaessa** [FOSSBilling/FOSSBilling#195]
We have an error code system, but it's not really being used. We should start using this system and build our documentation around it.

Ideally, each known error message should have a link attached to it and a documentation page made for it.

The documentation page should briefly explain the error, why it happens and some ways to solve it.

We could also set up short links for those documentation pages, like fossbilling.org/e/9876.

This would save us a lot of time for common errors. We wouldn't have to assist each user individually.

**Comment by evrifaessa** [comment]
I'll create some mockups about this and possibly start using this system for some of our most common errors. Any suggestions?

**Comment by jaapmarcus** [comment]
Start with 
0xxx = Core  errors
2xxx = server module 
3xxx = domain module
And so on higher the number further away from the core. 

If we think we might need more numbers of core reserve 0000 - 1999  or even 2999

**Comment by John-S4** [comment]
> We have an error code system,

Do we? 

I took a look through all of the bb documentation and couldn't find any mention of what the existing error codes mean - have you seen it anywhere, or do I need to look for it in code?

**Comment by evrifaessa** [comment]
> Do we?

https://github.com/FOSSBilling/FOSSBilling/blob/315c526f4a69553bc5f360db5fc9682b151cc386/src/bb-library/Box/Exception.php#L23-L31

The exception class accepts an optional code for the error.

**Comment by evrifaessa** [comment]
In some places, there are error codes already assigned. I don't know how we'll handle these. I think we could start over and give them new error codes.

https://github.com/FOSSBilling/FOSSBilling/blob/315c526f4a69553bc5f360db5fc9682b151cc386/src/bb-modules/Api/Controller/Client.php#L226

**Comment by evrifaessa** [comment]
I also think we should redesign the error page. It's the one grabbed from BoxBilling. I think the new layout should give more importance to the error code and the related documentation page.

**Comment by John-S4** [comment]
> I also think we should redesign the error page. It's the one grabbed from BoxBilling. I think the new layout should give more importance to the error code and the related documentation page.

Agreed. They are not really helpful at all the way that they are currently.

**Comment by John-S4** [comment]
> In some places, there are error codes already assigned. I don't know how we'll handle these. I think we could start over and give them new error codes.

Yes, just forget about them and start with a new numbering system instead of trying to incorporate any existing ones.

**Comment by Anuril** [comment]
I've looked over the error codes that are currently existing, and prepared a page. you can see it here: 

https://github.com/Anuril/fossbilling.org/blob/main/pages/docs/under-the-hood/error-handling.mdx

This is just the current state, not how the future should look.

**Comment by Ron-Zarbyte** [comment]
Looked through in an attempt to offer some feedback.

Would each error code be reserved for a specific place in code, or would similar errors use the same error code in different places? For example:

```
|  706 | Registrar related errors |
```

It may make the scope of the documentation page a bit too large depending on exactly how many error conditions would share the same codes.

Apart from that the idea sounds good. I like the reservation of code ranges for specific systems.

It may be good to have a master reference of error codes within code somewhere. For example some file that relates each error code to its label (using translations). Having some file that gets updated in repo each time a new error code is added may help make sure two or more people do not implement the same unused error code in future changes. Something like a basic helper class would work.

If you did intend to make multiple error messages for some codes then that idea may not work.

If there is anything I can help with on this one let me know.

**Comment by BelleNottelling** [comment]
> Would each error code be reserved for a specific place in code, or would similar errors use the same error code in different places? For example:

Probably a mix of both (in my opinion).
For more specific errors / exceptions where we can document fixes for end-users, my idea was actually that we could do translatable & helpful messages that are specific to that error code, including links that can take the user directly to documentation that is specific to that error code.

The list of that is here (there's only a few that I added when the error page got reworked):
https://github.com/FOSSBilling/FOSSBilling/blob/main/src/library/FOSSBilling/ErrorPage.php#L38

For everything else, *probably* mostly just somewhat generic error codes. I don't think it's super feasible to have a separate error code for every individual exception we might throw in the code and in most cases and that doing so would be overkill. 

Honestly, it's something that we have discussed on the Discord numerous times and never come to a conclusion for, everyone has differing opinions.

**Comment by Ron-Zarbyte** [comment]
Seems good to me and I agree, definitely a big step in the right direction.

That new `FOSSBillingErrorPage::getCodes()` is exactly what I was thinking. Next time I should actually look at the WIP code...

Codes that get used in more than one place that are later found to have differing diag steps can easily be changed to new codes anyway. So really it is a non-issue and was mostly for me being sure I understood more of future design plan.

I personally do not mind the extra legwork to throw more unique error codes in more places. That can definitely be debated endlessly. But it is also not really an issue since this new format would support both shared and unique codes.

I would agree and say go ahead and proceed on what has been done so far. I do like the error code format. It handles translations and it will be simple to maintain. The line count will grow over time but I think its a fine tradeoff for maintainability and keeping them in version control for easy conflict check. I could not think of a use case this format will be a problem for.

**Comment by jstebenne** [comment]
I just randomly got here and have very little context.

Why is this necessary or wanted? From my personal experience, handling error codes is more obscure than handling error instances. As a user or admin, they cause more confusion than is often required.

Why not instead simply display the error message as it is generated instead of adding a second layer? This would not remove the ability to include a link to the relevant documentation page (each error class can contain its own documentation url), and it would remove the burden of assigning error codes in a way that makes sense, especially with an evolving codebase.

**Comment by Anuril** [comment]
> I just randomly got here and have very little context.
There's quite a lot of errors where incomplete, weird or straight names and numbers are used, on multiple occasions, error messages have been copied and how the same error code and message is used for completely different things... 

> Why is this necessary or wanted? 
Not 100% sure, but I was frustrated with some annoying duplicates when I debugged something and then got a bit too into the idea at some point. 

> From my personal experience, handling error codes is more obscure than handling error instances. As a user or admin, they cause more confusion than is often required.
True... but "Could not save transaction id in handlePmntGatewayTx" doesn't help either. (Just an example) 

> Why not instead simply display the error message as it is generated instead of adding a second layer? 
Because the same error can happen in different parts of the code, but mean different things.

> This would not remove the ability to include a link to the relevant documentation page (each error class can contain its own documentation url), 
If the URL to the documentation is in the code, this will inevitably lead to broken links in error messages. Also, if the documentation needs restructuring, having a redirect from fossbilling.org/docs/kb/123 to docs.fossbilling.org/kb/123 is easier than to find which url has been used in what error message

> and it would remove the burden of assigning error codes in a way that makes sense, especially with an evolving codebase.
That is indeed a burden and to be fair the solution isn't where it should be yet.

![yza8p5b4ig661](https://github.com/user-attachments/assets/c8132cca-7177-4d71-9ec1-c2a1a1fadb85)
